README.md
# fusegen

Tools for generating and monitoring a Linux FUSE user-space file system.

## fuseGen

FuseGen is a Python 3 script which generates the basis of a FUSE file system
in the form
of a mix of Autotools configuration files and ANSI C source code.  The
C files constitute a **pass-through** FUSE file system.  This means that FUSE
operations are logged and statistics gathered, and then the commands are
passed through to the local file system for execution.

### FuseGen Command Line

	usage: fuseGen [-h] [-A ACPREREQ] [-D MYDATE] [-E EMAILADDR] [-I] [-f] [-j]
	               [-P PKGNAME] [-T] [-v] [-V MYVERSION]
	
	generate a minimal FUSE (file system in user space) package
	
	optional arguments:
	  -h, --help            show this help message and exit
	  -A ACPREREQ, --acPrereq ACPREREQ
	                        prerequisite autoconfig version number
	  -D MYDATE, --myDate MYDATE
	                        date in YYYY-MM-DD format
	  -E EMAILADDR, --emailAddr EMAILADDR
	                        contact email address
	  -I, --instrumenting   generate instrumentation code
	  -f, --force           if utility already exists, overwrite it
	  -j, --justShow        show options and exit
	  -P PKGNAME, --pkgName PKGNAME
	                        utility package name
	  -T, --testing         this is a test run
	  -v, --verbose         be chatty
	  -V MYVERSION, --myVersion MYVERSION
	                        version in X.Y.Z format

This is the command line at the time of writing; when you unpack
fusegen you should type `fuseGen -h` to see the current set of
supported arguments.

### Typical FuseGen Command Line: genEm

When fusegen is unpacked, the distribution directory contains a file
`genEm`.  As it stands this creates the `xxxfs` file system.  That is,
it writes a number of files to a subdirectory of your development
directory.

	cd ~/dev/py/fusegen
	./fuseGen -fIvP xxxfs

In this case, the package name is **xxxfs** (because the follows the **-P**),
and that will be the name
of the development directory.  If the directory already exists, it is
overwritten (**-f**).  Instrumentation code will be generated (**-I**)

### File System Created

`genEm` creates a number of files and directories.   What you see below
is a snapshot taken shortly after initialization (in fact, after running
the **build** command but without the object files and executables
generated my `make`).

	/home/jdd/dev/c/xxxfs
	├── aclocal.m4
	├── AUTHORS
	├── autogen.sh
	├── autom4te.cache
	├── bin
	│   ├── blk-31-4k
	│   ├── mountXXXFS
	│   └── umountXXXFS
	├── build
	├── ChangeLog
	├── CHANGES
	├── config
	├── config.guess
	├── config.log
	├── config.status
	├── config.sub
	├── configure
	├── configure.ac
	├── COPYING
	├── COPYING.AUTOCONF.EXCEPTION
	├── COPYING.GNUBL
	├── COPYING.LIB
	├── doc
	├── examples
	├── INSTALL
	├── libtool
	├── m4
	├── Makefile
	├── Makefile.am
	├── Makefile.in
	├── man
	├── NEWS
	├── README
	├── README.licenses
	├── scripts
	├── src
	│   ├── access.inc
	│   ├── bmap.inc
	│   ├── chmod.inc
	│   ├── chown.inc
	│   ├── config.h
	│   ├── config.h.in
	│   ├── create.inc
	│   ├── destroy.inc
	│   ├── fallocate.inc
	│   ├── fgetattr.inc
	│   ├── flock.inc
	│   ├── flush.inc
	│   ├── fsyncdir.inc
	│   ├── fsync.inc
	│   ├── ftruncate.inc
	│   ├── fuse_common.h
	│   ├── fuse.h
	│   ├── fuse_opt.h
	│   ├── fuse_version.h
	│   ├── getattr.inc
	│   ├── getxattr.inc
	│   ├── init.inc
	│   ├── link.inc
	│   ├── listxattr.inc
	│   ├── lock.inc
	│   ├── main.inc
	│   ├── Makefile
	│   ├── Makefile.am
	│   ├── Makefile.in
	│   ├── mkdir.inc
	│   ├── mknod.inc
	│   ├── opcodes.h
	│   ├── opendir.inc
	│   ├── open.inc
	│   ├── optable.inc
	│   ├── readdir.inc
	│   ├── read.inc
	│   ├── readlink.inc
	│   ├── releasedir.inc
	│   ├── release.inc
	│   ├── removexattr.inc
	│   ├── rename.inc
	│   ├── rmdir.inc
	│   ├── setxattr.inc
	│   ├── stamp-h1
	│   ├── statfs.inc
	│   ├── symlink.inc
	│   ├── truncate.inc
	│   ├── unlink.inc
	│   ├── util.c
	│   ├── utimens.inc
	│   ├── write.inc
	│   ├── xxxfs
	│   ├── xxxfs.c
	│   ├── xxxfs.h
	├── tmp
	│   └── bucket-20150316-150838
	├── workdir
	│   ├── mountPoint
	│   └── rootdir
	│       ├── foo
	│       ├── job1.1.0
	│       ├── job2.2.0
	│       ├── job3.3.0
	│       └── job4.4.0
	└── xxxfs.log

For easy customization there is a `.inc` file for each FUSE command.

### The build Command

The `build` command completes the creation of the file system.  It runs
`autogen.sh`, which installs many of the standard Autotools files, and
then creates the `configure` script from the instructions in `configure.ac`.
Finaly it runs make, which compiles the object files and then links the
file system, the `xxxfs` executable.

	./autogen.sh
	./configure
	make

It may be desirable to change the generated system to some degree.
Usually limited changes are accomplished by editing `configure.ac`
and/or `Makefile.ac` and then rerunning build.  This will have no
effect on `src/*.inc` and in that sense is safe.

### bin/ Commands

The `bin/` subdirectory contains scripts for

* mounting the file system (`mountXXXFS` in this case)
* running a simple FIO benchmark (`blk-31-4k`)
* and dismounting the file system (`umountXXXFS`)

## Putting It All Together: Running an Application

The script `blk-31-4k`, found in the default `bin/` directory, is used
to collect statistics on a short **fio** run.  FIO is an open source tool
for measuring the performance of a file system.

	echo "blk-31-4k: test size is being set to $1 MB"
	#
	cd ~/dev/c/xxxfs
	bin/mountXXXFS
	cd workdir/mountPoint
	fio --name=global --bs=4k --size=$1m 	--rw=randrw --rwmixread=75 	--name=job1 --name=job2 --name=job3 --name=job4
	cd ../..
	bin/umountXXXFS

This particular script runs four jobs simultaneously.  Each job reads
and/or writes the same amount of data to the disk.  On the command line
the script has a single argument, the number of megabytes to be read from
and/or written to the disk.  For this test the mount point and the root
directory are both subdirecties of `workdir`.  The script changes to the
directory above that, then mounts the FUSE file system (`bin/mountXXXFS`).
It then runs the FIO jobs in the `mountPoint` subdirectory.  When the FIO
jobs terminate, the script changes back to the distribution directory and
unmounts the file system.

Running this script has two major effects:

* it writes its log to the file `xxxfs.log`
* it writes run statistics to a file under `tmp/`

If full logging is turned on, the log file can be very large.

## FuseGen Statistics File

The statistics file has a name like `bucket-20150316-150838`.  The first
set of numbers in the file name represents the date in `CCYYMMDD` format; the
second set is the start time of the run, local time, as `HHMMSS`.  We use
a 24-hour clock, so in this case the run started at 15:08:38, just
after 3 pm local time.

	typedef struct o_ {
	    uint32_t    opSec;
	    uint32_t    opNsec;             // may not exceed BILLION
	    uint32_t    lateNsec;           // ns of latency; may not exceed BILLION
	    unsigned    lateSec     :  7;   // sec of latency
	    unsigned    count       : 17;   // bytes read or written; 128K max
	    unsigned    opCode      :  8;
	} __attribute__((aligned(16), packed)) opData_t;

Each call on the FUSE file system is recorded to the statistics file as
a 16-byte object laid out as specified above.

The time of the call is recorded as a 128-bit pair, `opSec` and
`opNsec`.  This is a Linux CLOCK_MONOTONIC time representing an absolute
elapsed time; it is generally **not** the same as the system's (guess at)
the local time.

When the system call is complete, the latency is calculated.  This has the
same 32-bit nanosecond part, but only 7 bits are reserved for the seconds
of latency.  That is, latencies greater than 128 seconds will not be recorded
correctly.

Finally 8 bits are reserved for the FUSE opcode.  There are currently
fewer than four dozen opcodes so this number of bits is likely to be
sufficient for some time.

## FuseDecode

FuseDecode is a utility for converting FuseGen statistics files to a
human-readable format.

### FuseDecode Command Line

	usage: fuseDecode [-h] [-D PATHTODATA] [-j] [-n HOWMANY] [-v]
	
	parse a fuseGenned data file
	
	optional arguments:
	  -h, --help            show this help message and exit
	  -D PATHTODATA, --pathToData PATHTODATA
	                        path to data file
	  -j, --justShow        show options and exit
	  -n HOWMANY, --howMany HOWMANY
	                        how many data points to decode
	  -v, --verbose         be chatty

### Typical FuseDecode Output

	==== op time === == latency == = byte = == opcode ==     = flags =
	  sec   nanosec  sec  nanosec   count   op   name
	369886 170431079  0     74955       0   29 init
	369886 170592946  0     91116       0    0 getattr
	369886 170601627  0     58834       0   25 opendir          P
	369886 170728424  0     68768       0   26 readdir
	369886 170849263  0     60370       0    0 getattr
	369886 170851394  0     80361       0    0 getattr          P
	369886 170937891  0     22188       0   31 access
	369886 170972000  0     52993       0    0 getattr
	369886 171057175  0     28741       0    0 getattr
	369886 171112174  0     21897       0    0 getattr
	369886 171181270  0     26367       0    0 getattr
	369886 171243684  0     22415       0    0 getattr
	369886 171301507  0     41395       0    0 getattr
	369886 171378102  0     24336       0    0 getattr
	369886 171483916  0     21530       0   27 releasedir
	369886 171505940  0     30953       0   25 opendir
	369886 171587944  0     47520       0   26 readdir
	369886 171748142  0     16430       0   27 releasedir
	369886 171768149  0     25347       0   25 opendir
	369886 171832503  0     46673       0   26 readdir
	369886 171986430  0     70618       0   27 releasedir
	369886 172007454  0     53362       0   25 opendir          P
	369886 172099167  0     49953       0   26 readdir

Output from a typical run of the `blk-31-4k` script is available
[here](runData-20150316-135941.gz)

## Project Status

A good beta.  Tests succeed.

## On-line Documentation

More information on the **fusegen** project can be found
[here](https://jddixon.github.io/fusegen)


CHANGES
~/dev/py/fusegen/CHANGES

v0.6.37
    2018-03-05
        * edit setup.py, tox config files                           SLOC 4581
v0.6.36
    2017-12-12
        * more pylinting plus more bug fixes                        SLOC 4560
v0.6.35
    2017-12-07
        * pep8 (needs more) plus minor bug fixes                    SLOC 4573 
v0.6.34
    2017-09-25
        * add projConfig.toml                                       SLOC 4568
v0.6.33
    2017-09-13
        * pep8 etc; correct tox.ini, requirements.txt               SLOC 4539
v0.6.32
    2017-09-08
        * pep8 plus add tox.ini, *req*.txt                          SLOC 4524
v0.6.31
    2017-07-16
        * rewrite setup.py, MANIFEST.in; edit cfg files             SLOC 4497
v0.6.30
    2017-06-10
        * restructure for pytest                                    SLOC 4376
v0.6.29
    2017-04-23
        * drop dist/* from repo; plus minor edits                   SLOC 4546
v0.6.28
    2016-12-07
        * pylinted to 9.08/10; switched to dump_options()           SLOC 4541
v0.6.27
    2016-10-17
        * add pylint, coverage; rename files per PEP 8              SLOC 4483
v0.6.26
    2016-08-24
        * fix license, setup.py                                     SLOC 4412
v0.6.25
    2016-06-21
        * add ghpDoc/{css,img}/                                     SLOC 4341
v0.6.24
    2016-05-20
        * add bkp2U, update .dvcz etc                               SLOC 4390
v0.6.23
    2016-04-26
        * added .dvcz/projCfg{,.local}                              SLOC 3668
v0.6.22
    2016-04-10
        * added .fileMap, dropped extern from count                 SLOC 3741
v0.6.21
    2016-02-24
        * add auto8it and execute                                   SLOC 4317
v0.6.20
    2016-01-27
        * minor edits                                               SLOC 4268
v0.6.19
    2015-12-20
        * fix dead links                                            SLOC 4268
v0.6.18
    2015-11-14
        * minor edits to *.md                                       SLOC 4268
v0.6.17
    2015-09-02
        * README and index.md synchronized                          SLOC 4266
v0.6.16
    2015-06-23
        * update .gitignore; minor cleanup; add installit           SLOC 4265
v0.6.15
    2015-04-11
        * simplified handling of ops not yet implemented
        * add -z option to fuseDecode, to select zero-count ops
        * dropped README.md from countem (253 sloc)                 SLOC 2159
v0.6.14
    2015-03-25
        * add -S/--stats fuseDecode option: opcodes used, counts    SLOC 2383
        * testFuseGeneration four way test                          SLOC 2383
v0.6.13
    2015-03-24
        * testFuseGeneration: trees match                           SLOC 2343
v0.6.12
    2015-03-23
        * testFuseGeneration writes random directory                SLOC 2331
v0.6.11
    2015-03-22
        * initial work on testFuseGeneration.py                     SLOC 2312
v0.6.10
    2015-03-21
        * genEm, genEmI, genEmL, genEmIL build correctly            SLOC 2230
v0.6.9
    2015-03-19
        * elaboration of README.md                                  SLOC 2073
    2015-03-18
        * add skeletal README.md                                    SLOC 2073
v0.6.8
    2015-03-16
        * opData_t: maxRead/Write limit 128KB, 2^7 sec latency      SLOC 1873
        * prevent segfaults by dynamically reallocating op table    SLOC 1894
v0.6.7
    2015-03-15
        * begin work on bmap                                        SLOC 1864
v0.6.6
    2015-03-14
        * add flock                                                 SLOC 1858
v0.6.5
    2015-03-13
        * add lock op                                               SLOC 1850
v0.6.4
    2015-03-11
        * add utimens, deprecate utime, fix fallocate               SLOC 1826
v0.6.3
    2015-03-10
        * regularized naming of functions, variables                SLOC 1798
v0.6.2
    2015-03-09
        * add gustav_ctries.c, majek_siphash.c (SLOC excludes *.c)  SLOC 1793
v0.6.1
    2015-03-08
        * licensing changes                                         SLOC 1793
v0.6.0
    2015-03-07
        * added opcodes.h, fixed testFuseFunc.py, created stdi_fs,
            executable depends upon *.inc                           SLOC 1793 
v0.5.9
    2015-03-06
        * fallocate implemented                                     SLOC 1589
        * generate bin/blk-31-4k                                    SLOC 1604
        * fuseDecode decodes data files                             SLOC 1731
v0.5.8
    2015-03-05
        * writeBucket() working                                     SLOC 1560
v0.5.7
    2015-03-04
        * add build, writeBucket(), misc fixes                      SLOC 1535
v0.5.6
    2015-03-03
        * generates clockMeIn/clockMeOut() calls                    SLOC 1499
v0.5.5
    2015-03-02
        * completed handling command line                           SLOC 1476
v0.5.4
    2015-03-01
        * command line arguments                                    SLOC 1394
v0.5.3
    2015-02-28
        * fiddling with data strucures                              SLOC 1356
v0.5.2
    2015-02-27
        * extended util.c                                           SLOC 1353
v0.5.1
    2015-02-26
        * re-rewrote util.c                                         SLOC 1310
v0.5.0
    2015-02-25
        * began adding instrumentation                              SLOC 1290
v0.4.13
    2015-02-24
        * rewrote util.c                                            SLOC 1265
v0.4.12
    2015-02-23
        * fixed bug in readdir                                      SLOC 1218
v0.4.11
    2015-02-22
        * improved checks on CLI arg list                           SLOC 1220
v0.4.10
    2015-02-21
        * generate util.c; more fixes                               SLOC 1207
v0.4.9
    2015-02-20
        * add missing FUSE header files; minor fixes                SLOC 1064
v0.4.8
    2015-02-19

test_fuse_generation.py
#!/usr/bin/env python3

# fusegen/testFuseGeneration.py

import base64
import hashlib
import os
import sys
import time
import unittest
from argparse import Namespace
from rnglib import SimpleRNG
from fusegen import invoke_shell, make_fuse_pkg, SH
from merkletree import MerkleTree


class TestFuseGeneration(unittest.TestCase):

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    # FUNCTIONS TO MODIFY IN-MEMORY DATA STRUCTURE AND DISK IMAGE
    # XXX STUB XXX

    # FUNCTIONS TO DETERMINE EQUALITY OF IN-MEMORY DATA STRUCTURE
    # AND DISK IMAGE (either under mountPoint or rootDir)
    # XXX STUB XXX

    def fiddle_with_files(self, pkg_name, path_to_pkg, umount_cmd):
        """ Enter with the file system mounted """
        work_dir = os.path.join(path_to_pkg, 'workdir')
        mount_point = os.path.join(work_dir, 'mount_point')
        root_dir = os.path.join(work_dir, 'rootdir')

        # Devise a directory structure, say M files wide, N directories deep.
        # The files are of random-ish length, populated with random-ish data.
        sample_name = self.rng.next_file_name(16)
        path_to_sample = os.path.join(mount_point, sample_name)
        # builds a directory tree with a depth of 4, 5 files (including
        # directories) at each level, and 16 <= file length <= 128
        self.rng.nextDataDir(path_to_dir=path_to_sample,
                             depth=4, width=5, max_len=128, min_len=16)
        # DEBUG
        print("creating tree1")
        sys.stdout.flush()
        # END
        tree1 = MerkleTree.create_from_file_system(path_to_sample)
        self.assertTrue(tree1 is not None)

        # If this succeeds, we have written the directory structure on the
        # mount point.

        # Delete some files, modifying in-memory data structure accordingly
        # Shorten some files, modifying in-memory data structure accordingly
        # Lengthen some files, modifying in-memory data structure accordingly

        # Unmount the file system
        chatter = invoke_shell(umount_cmd)

        # Verify that the expected directory structure appears below
        # the root directory.
        path_via_root = os.path.join(root_dir, sample_name)
        # DEBUG
        print("creating tree2")
        sys.stdout.flush()
        # END
        tree2 = MerkleTree.create_from_file_system(path_via_root)
        self.assertTrue(tree2 is not None)
        self.assertEqual(tree1.equal(tree2), True)
        # DEBUG
        print("directory trees are equal")
        sys.stdout.flush()
        # END
        return chatter

    def exercise_file_system(self, pkg_name, path_to_pkg):
        dir_now = os.getcwd()
        os.chdir(path_to_pkg)
        path_to_bin = os.path.join(path_to_pkg, 'bin')

        mount_cmd = [
            SH,
            os.path.join(
                path_to_bin,
                'mount%s' %
                pkg_name.upper()),
        ]
        umount_cmd = [
            SH,
            os.path.join(
                path_to_bin,
                'umount%s' %
                pkg_name.upper()),
        ]

        chatter = ''
        try:
            chatter = invoke_shell(mount_cmd)
            chatter += self.fiddle_with_files(pkg_name,
                                              path_to_pkg, umount_cmd)
        except Exception as exc:
            print(exc)

        else:
            # XXX STUB XXX
            pass

        finally:
            """ unmount the file system, ignoring any exceptions """
            # DEBUG
            print("enter finally block")
            sys.stdout.flush()
            # END
            try:
                invoke_shell(umount_cmd)
            except:
                pass

            if chatter and chatter != '':
                print(chatter)
            os.chdir(dir_now)

            # DEBUG
            print("after fiddling with files we are back in %s" % dir_now)
            sys.stdout.flush()
            # END

    def do_bae_test(self, logging=False, instrumenting=False):
        """
        Build the selected type of file system under devDir and
        then run exerciseFileSystem() on it.
        """
        dev_dir = '/home/jdd/dev/c'
        pkg_name = 'xxxfs'
        if instrumenting:
            pkg_name += 'I'
        if logging:
            pkg_name += 'L'
        path_to_pkg = os.path.join(dev_dir, pkg_name)

        cmds = Namespace()
        setattr(cmds, 'ac_prereq', '2.6.9')
        setattr(cmds, 'dev_dir', dev_dir)
        setattr(cmds, 'email_addr', 'jddixon at gmail dot com')
        setattr(cmds, 'force', True)
        setattr(cmds, 'instrumenting', instrumenting)
        setattr(cmds, 'logging', logging)
        setattr(cmds, 'my_date', "%04d-%02d-%02d" % time.gmtime()[:3])
        setattr(cmds, 'my_version', '1.2.3')
        setattr(cmds, 'path_to_pkg', path_to_pkg)
        setattr(cmds, 'pkg_name', pkg_name)
        setattr(cmds, 'lc_name', pkg_name.lower())
        setattr(cmds, 'uc_name', pkg_name.upper())
        setattr(cmds, 'testing', False)
        setattr(cmds, 'verbose', False)

        # DEBUG
        print(cmds)
        # END

        # create the target file system
        make_fuse_pkg(cmds)

        # invoke the build command
        dir_now = os.getcwd()
        os.chdir(path_to_pkg)
        cmd = [SH, os.path.join(path_to_pkg, 'build'), ]
        chatter = ''
        try:
            chatter = invoke_shell(cmd)
        except Exception as exc:
            print(exc)
        if chatter and chatter != '':
            print(chatter)
        os.chdir(dir_now)

        # DEBUG
        print("we are back in %s" % dir_now)
        # END

        # run test verifying that the file system works as expected
        self.exercise_file_system(pkg_name, path_to_pkg)

    def do_instruments_test(self):
        self.do_bae_test(instrumenting=True)

    def do_logging_test(self):
        self.do_bae_test(logging=True)

    def do_test_logging_and_instrumented(self):
        self.do_bae_test(logging=True, instrumenting=True)

    def test_fuse_generation(self):
        self.do_bae_test()
        self.do_instruments_test()
        self.do_logging_test()
        self.do_test_logging_and_instrumented()

if __name__ == '__main__':
    unittest.main()


tests/test_fuse_func.py
#!/usr/bin/env python3

# testFuseFunc.py

""" Test various fuse-related functions. """

import unittest

# from rnglib        import SimpleRNG
from fusegen import FuseFunc


class TestFuseFunc(unittest.TestCase):
    """ Test various fuse-related functions. """

    def setUp(self):
        pass

    def tearDown(self):
        pass

    # utility functions #############################################

    # actual unit tests #############################################

    def test_name_to_func_map(self):
        """ Verify map from function names to opcodes is correct. """

        match_, o_map = FuseFunc.get_func_map('xxx_')   # funcMap, opCodeMap
        _ = o_map               # suppress warning
        # for testing, build a new p2tMap from the data in funcMap
        for name in match_:
            func = match_[name]
            params = func.params        # a list of 2-tuples
            my_p2t = {}    # maps parameter names to type as a string
            p2t_map = func.p2t_map
            for param_ in params:
                p_type = param_[0]
                p_name = param_[1]
                my_p2t[p_name] = p_type

            # now check the maps for equality
            self.assertEqual(len(my_p2t), len(p2t_map))
            for param_ in p2t_map:
                p_type_name = p2t_map[param_]   # param type from param name
                p_type = my_p2t[param_]
                self.assertEqual(p_type_name, p_type)

    def test_first_line(self):
        """ UNFINISHED: test handling of first line. """

        match_, o_map = FuseFunc.get_func_map('xxx_')  # funcMap, opCodeMap
        _ = o_map               # suppress warning
        _ = self                # suppress warning
        for name in match_:
            func = match_[name]
            line = func.first_line()
            # DEBUG
            print(line)
            # END

            # TODO: compare generated line with corresponding line
            # in the file PATH_TO_FIRST_LINES


if __name__ == '__main__':
    unittest.main()


testFuseFunc.py
#!/usr/bin/env python3

# testFuseFunc.py

import os
import sys
import unittest

#from rnglib        import SimpleRNG
from fusegen import FuseFunc, OP_NAMES, PATH_TO_FIRST_LINES


class TestFuseFunc (unittest.TestCase):

    def setUp(self):
        pass

    def tearDown(self):
        pass

    # utility functions #############################################

    # actual unit tests #############################################

    def testNameToFuncMap(self):
        m, o = FuseFunc.getFuncMap('xxx_')   # funcMap, opCodeMap
        # for testing, build a new p2tMap from the data in funcMap
        for name in m:
            func = m[name]
            params = func.params        # a list of 2-tuples
            myP2T = {}    # maps parameter names to type as a string
            p2tMap = func.p2tMap
            for p in params:
                pType = p[0]
                pName = p[1]
                myP2T[pName] = pType

            # now check the maps for equality
            self.assertEqual(len(myP2T), len(p2tMap))
            for p in p2tMap:
                t = p2tMap[p]   # param type from param name
                T = myP2T[p]
                self.assertEqual(t, T)

    def testFirstLine(self):
        m, o = FuseFunc.getFuncMap('xxx_')  # funcMap, opCodeMap
        for name in m:
            func = m[name]
            line = func.firstLine()
            # DEBUG
            print(line)
            # END

            # TODO: compare generated line with corresponding line
            # in the file PATH_TO_FIRST_LINES

if __name__ == '__main__':
    unittest.main()


testFuseGeneration.py
#!/usr/bin/env python3

# fusegen/testFuseGeneration.py

import base64
import hashlib
import os
import sys
import time
import unittest
from argparse import Namespace
from rnglib import SimpleRNG
from fusegen import invokeShell, makeFusePkg, SH
from merkletree import *


class TestFuseGeneration (unittest.TestCase):

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    # FUNCTIONS TO MODIFY IN-MEMORY DATA STRUCTURE AND DISK IMAGE
    # XXX STUB XXX

    # FUNCTIONS TO DETERMINE EQUALITY OF IN-MEMORY DATA STRUCTURE
    # AND DISK IMAGE (either under mountPoint or rootDir)
    # XXX STUB XXX

    def fiddleWithFiles(self, pkgName, pathToPkg, umountCmd):
        """ Enter with the file system mounted """
        workDir = os.path.join(pathToPkg, 'workdir')
        mountPoint = os.path.join(workDir, 'mountPoint')
        rootDir = os.path.join(workDir, 'rootdir')

        # Devise a directory structure, say M files wide, N directories deep.
        # The files are of random-ish length, populated with random-ish data.
        sampleName = self.rng.nextFileName(16)
        pathToSample = os.path.join(mountPoint, sampleName)
        # builds a directory tree with a depth of 4, 5 files (including
        # directories) at each level, and 16 <= file length <= 128
        self.rng.nextDataDir(pathToDir=pathToSample,
                             depth=4, width=5, maxLen=128, minLen=16)
        # DEBUG
        print("creating tree1")
        sys.stdout.flush()
        # END
        tree1 = MerkleTree.createFromFileSystem(pathToSample)
        self.assertTrue(tree1 is not None)

        # If this succeeds, we have written the directory structure on the
        # mount point.

        # Delete some files, modifying in-memory data structure accordingly
        # Shorten some files, modifying in-memory data structure accordingly
        # Lengthen some files, modifying in-memory data structure accordingly

        # Unmount the file system
        chatter = invokeShell(umountCmd)

        # Verify that the expected directory structure appears below
        # the root directory.
        pathViaRoot = os.path.join(rootDir, sampleName)
        # DEBUG
        print("creating tree2")
        sys.stdout.flush()
        # END
        tree2 = MerkleTree.createFromFileSystem(pathViaRoot)
        self.assertTrue(tree2 is not None)
        self.assertEqual(tree1.equal(tree2), True)
        # DEBUG
        print("directory trees are equal")
        sys.stdout.flush()
        # END
        return chatter

    def exerciseFileSystem(self, pkgName, pathToPkg):
        dirNow = os.getcwd()
        os.chdir(pathToPkg)
        pathToBin = os.path.join(pathToPkg, 'bin')

        mountCmd = [SH, os.path.join(pathToBin, 'mount%s' % pkgName.upper()), ]
        umountCmd = [
            SH,
            os.path.join(
                pathToBin,
                'umount%s' %
                pkgName.upper()),
        ]

        chatter = ''
        try:
            chatter = invokeShell(mountCmd)
            chatter += self.fiddleWithFiles(pkgName, pathToPkg, umountCmd)
        except Exception as e:
            print(e)

        else:
            # XXX STUB XXX
            pass

        finally:
            """ unmount the file system, ignoring any exceptions """
            # DEBUG
            print("enter finally block")
            sys.stdout.flush()
            # END
            try:
                invokeShell(umountCmd)
            except:
                pass

            if chatter and chatter != '':
                print(chatter)
            os.chdir(dirNow)

            # DEBUG
            print("after fiddling with files we are back in %s" % dirNow)
            sys.stdout.flush()
            # END

    def doBaseTest(self, logging=False, instrumenting=False):
        """
        Build the selected type of file system under devDir and
        then run exerciseFileSystem() on it.
        """
        devDir = '/home/jdd/dev/c'
        pkgName = 'xxxfs'
        if instrumenting:
            pkgName += 'I'
        if logging:
            pkgName += 'L'
        pathToPkg = os.path.join(devDir, pkgName)

        cmds = Namespace()
        setattr(cmds, 'acPrereq', '2.6.9')
        setattr(cmds, 'devDir', devDir)
        setattr(cmds, 'emailAddr', 'jddixon at gmail dot com')
        setattr(cmds, 'force', True)
        setattr(cmds, 'instrumenting', instrumenting)
        setattr(cmds, 'logging', logging)
        setattr(cmds, 'myDate', "%04d-%02d-%02d" % time.gmtime()[:3])
        setattr(cmds, 'myVersion', '1.2.3')
        setattr(cmds, 'pathToPkg', pathToPkg)
        setattr(cmds, 'pkgName', pkgName)
        setattr(cmds, 'lcName', pkgName.lower())
        setattr(cmds, 'ucName', pkgName.upper())
        setattr(cmds, 'testing', False)
        setattr(cmds, 'verbose', False)

        # DEBUG
        print(cmds)
        # END

        # create the target file system
        makeFusePkg(cmds)

        # invoke the build command
        dirNow = os.getcwd()
        os.chdir(pathToPkg)
        cmd = [SH, os.path.join(pathToPkg, 'build'), ]
        chatter = ''
        try:
            chatter = invokeShell(cmd)
        except Exception as e:
            print(e)
        if chatter and chatter != '':
            print(chatter)
        os.chdir(dirNow)

        # DEBUG
        print("we are back in %s" % dirNow)
        # END

        # run test verifying that the file system works as expected
        self.exerciseFileSystem(pkgName, pathToPkg)

    def doInstrumentedTest(self):
        self.doBaseTest(instrumenting=True)

    def doLoggingTest(self):
        self.doBaseTest(logging=True)

    def doTestLoggingAndInstrumented(self):
        self.doBaseTest(logging=True, instrumenting=True)

    def testFuseGeneration(self):
        self.doBaseTest()
        self.doInstrumentedTest()
        self.doLoggingTest()
        self.doTestLoggingAndInstrumented()

if __name__ == '__main__':
    unittest.main()
